Initial specs migration.

adlerjohn opened the pull request:
Fixes #792

An initial migration of the existing specs as of [`e59efd63a2165866584833e91e1cb8a6ed8c8203`](https://github.com/celestiaorg/celestia-specs/tree/e59efd63a2165866584833e91e1cb8a6ed8c8203). It's expected that there are a few broken links and extra content that has to be pruned, along with incorrect content. The important thing is to not miss anything.

Also sets up a CI workflow that will build the book.

rootulp commented on specs/src/specs/data_structures.md:
```diff
@@ -0,0 +1,994 @@
+| `Amount`                    | `uint64`                    |
+| `Graffiti`                  | `byte[MAX_GRAFFITI_BYTES]`  |
+| [`HashDigest`](#hashdigest) | `byte[32]`                  |
+| `Height`                    | `int64`                     |
```
[no change needed] I'm not sure why height is a signed integer rather than unsigned but this conforms to tendermint/tendermint ([ref](https://github.com/tendermint/tendermint/blob/aa303edaefeb8a0856d7494cca24e51f6f3e2393/state/errors.go#L13)) so no change needed

adlerjohn replied:
(Invalid) reasoning here: https://blog.cosmos.network/choosing-a-type-for-blockchain-height-beware-of-unsigned-integers-714804dddf1d

rootulp commented on specs/theme/highlight.js:
```diff
@@ -0,0 +1,1328 @@
+/*
+  Highlight.js 10.7.2 (00233d63)
```
[no change needed] this file looks :suspect: but it seems intentional
1. https://rust-lang.github.io/mdBook/format/theme/syntax-highlighting.html
2. https://highlightjs.org/
3. https://github.com/celestiaorg/celestia-specs/blob/master/theme/highlight.js

adlerjohn replied:
It's copy-pasted from the original repo without change.

rootulp commented on specs/README.md:
```diff
@@ -0,0 +1,22 @@
+
+## Building From Source
+
+To build book:
```
[optional]

```suggestion
Prerequisite: [install mdBook](https://rust-lang.github.io/mdBook/guide/installation.html)

To build book:
```

evan-forbes replied:
this is a good addition, but I'd prefer not to add it in this PR

adlerjohn replied:
Fixed in https://github.com/celestiaorg/celestia-app/pull/800

rootulp reviewed (dismissed):
LGTM! Thanks for your hard work on this

evan-forbes reviewed (approved):
dope, glad these are finding a home.

evan-forbes commented on specs/src/specs/consensus.md:
```diff
@@ -0,0 +1,770 @@
+| `TRANSACTION_NAMESPACE_ID`              | `NamespaceID` | `0x0000000000000001` | Transactions: requests that modify the state.                                              |
+| `INTERMEDIATE_STATE_ROOT_NAMESPACE_ID`  | `NamespaceID` | `0x0000000000000002` | Intermediate state roots, committed after every transaction.                               |
+| `EVIDENCE_NAMESPACE_ID`                 | `NamespaceID` | `0x0000000000000003` | Evidence: fraud proofs or other proof of slashable action.                                 |
+| `TAIL_TRANSACTION_PADDING_NAMESPACE_ID` | `NamespaceID` | `0x00000000000000FF` | Tail padding for transactions: padding after all transactions but before messages.         |
```
same as https://github.com/celestiaorg/celestia-app/pull/793#discussion_r980619637

liamsi commented on specs/src/specs/proto/wire.proto:
```diff
@@ -0,0 +1,98 @@
+  uint64 nonce = 2;
+}
+
+message WireTxBurn {
+  uint64 amount = 1;
+  TransactionFee fee = 2;
+  uint64 nonce = 3;
+  // 32-byte graffiti
+  bytes graffiti = 4;
+}
```
This Tx type does not exist. Neither any "graffiti" field. 
I am not sure that copying the spec as is without going through the hard work to update it is the best approach but up to you and the core/app team.

liamsi reviewed (commented):
I would ask to add John, Matt, and me to the `specs` sub-dir as code-owner.

liamsi commented on specs/src/specs/proto/types.proto:
```diff
@@ -0,0 +1,285 @@
+}
+// ANCHOR_END: BadEncodingFraudProof
+
+// ANCHOR: StateFraudProof
+// https://github.com/celestiaorg/celestia-specs/blob/master/specs/networking.md#statefraudproof
+message StateFraudProof {
```
Can be deleted I assume?

liamsi reviewed (commented):
> Also sets up a CI workflow that will push the built book to a main directory on the gh-pages branch on each merge to main, allowing a rendered form of the book to be displayed on GitHub pages.

I would recommend to remove this until the specs are somewhat up-to-date again. Otherwise, we will be confusing a bunch of folks visiting the published specs again.

musalbas commented:
Shouldn't the parts about how data is encoded be in Celestia core, not app? We want to have specs be in the repo that they're relevant for.

musalbas commented:
Also, it seems a large part of the spec is outdated (sparse Merkle trees, graffiti, etc). IMO there's not much advantage to merging the existing outdated specs into app, as people can already reference the existing specs repo. I think there should be a draft PR to do a basic update and clean up of the specs, as well as decoupling the bits for core and app, so that we have a general structure that can be built on, before merging this.

evan-forbes commented:
> I think there should be a draft PR to do a basic update and clean up of the specs, as well as decoupling the bits for core and app, so that we have a general structure that can be built on, before merging this.

@musalbas I think there was a miscommunication, where those of us on the call agreed to move the specs over first, and then edit them with individual PRs. this is why Rootul and myself approved so quickly and why John made the PR in the first place.

if we don't want to do that, then we can turn this PR into a draft and make the individual PRs on this branch until we're comfortable with merging.

> Shouldn't the parts about how data is encoded be in Celestia core, not app?

all encoding code is kept in the application now, so I think it makes sense to keep these specs here.

> I would recommend to remove this until the specs are somewhat up-to-date again. Otherwise, we will be confusing a bunch of folks visiting the published specs again.

not publishing the specs until they match the implementation makes a lot of sense. We can split this portion into a different PR.

evan-forbes reviewed (changes requested):
we should split the CI publishing mechanism into a different PR https://github.com/celestiaorg/celestia-app/pull/793#pullrequestreview-1121612747

musalbas commented:
Makes sense, would it make sense to merge this into a different branch like `specs` in that case, until it's ready for `main`?

adlerjohn commented on specs/src/specs/block_proposer.md:
```diff
@@ -0,0 +1,26 @@
+
+Note: the maximum padding shares between messages should be at most twice the number of message shares. Doubling the square size (i.e. quadrupling the number of shares in the square) should thus only have to happen at most once.
+
+## Laying out Transactions and Messages
```
We do!

adlerjohn commented:
1. Made a new branch `specs-staging` that we can squash-merge PRs into, starting with this one, and re-targeted this PR on that branch. This avoids end-users seeing a WIP `specs` directory.
2. Removed the deploy step from CI, added a build step to ensure the book actually builds properly.

evan-forbes commented:
Thanks @adlerjohn! gets my :+1: :+1:
